raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -341,37 +356,59 @@ private VoteResponseData handleVoteRequest(
             return buildVoteResponse(Errors.INVALID_REQUEST, false);
         }
 
+        // we already confirmed that candidateEpoch >= quorum.epoch()
+        final boolean voteGranted;
         if (candidateEpoch > quorum.epoch()) {
-            // If the epoch is larger than our own, then we need to bump our own epoch,
-            // but we leave the vote undecided until we check the offset below
-            becomeUnattachedFollower(candidateEpoch);
```

Comment:
I know you didn't like this, but the logic was more concise. Now we have two separate paths where a vote could be granted instead of one.

Reply:
Actually the old logic has a flaw such that if it was a candidate or voter, it would then become unattached follower if epoch is larger but the end-offset validation fails. However the `becomeUnattachedFollower` is only needed to bump up the epoch to the newer value.

In unattached follower state, we would reset timer to the longer fetch timeout and send find-quorum afterwards. The correct behavior should be stay as candidate with the timer reset to election timeout.

Reply:
I think I'm missing something. With this patch anyway, we reset to the election timeout (plus jitter) in `becomeUnattachedFollower`. Is that not what we want?

Reply:
If we were previously a fetching voter, then the timer should be set for fetch timeout (which would be longer than election), and then upon becoming an unattached follower, if we reset it to `election timeout + jitter`, we are effectively shortening the time for this follower to eventually become a candidate. But if we were a fetching observer, we should not reset any timer.

I'm pondering that for the first case, should we actually shorten the timer from remaining fetch timeout to election timeout.